### DISCUSSION TITLE
Techempower benchmark spends 9% of time in System.nanoTime()?

### DISCUSSION THREAD

**Discussion by jfrantzius** [quarkusio/quarkus#30437] [Q&A] [answered]
Hi @franz1981 , following up to our previous [discussion on profiling the Techempower benchmark](https://github.com/quarkusio/quarkus/discussions/30231#discussioncomment-4715218), this is number 2 of the hotspots that JProfiler sees with async profiling (running the benchmark using Docker on x86 Linux), inbetween sending and receiving data, i.e. more time spent there than IO for reading data:

![image](https://user-images.githubusercontent.com/1823661/213130967-ef82da23-d2a7-4fe8-8b48-572c26c756aa.png)

By comparing the calling Java code, this seem to be calls to `System.nanoTime()`. Again it's suprising that this should be expensive, given that `clock_gettime()` resides in [vDSO](https://man7.org/linux/man-pages/man7/vdso.7.html) and so should be running in user space. But then JProfiler doesn't report the number of invocations.

The JVM's code seems to be this:
![image](https://user-images.githubusercontent.com/1823661/213133347-bc5995e9-69c6-4c72-84c3-61098e9098a5.png)

Do you have an idea what we are seeing here?

**Comment by jfrantzius** [comment]
This is what one call path into `System.nanoTime()` looks like in async-profiler (using `-e cycles`):

![image](https://user-images.githubusercontent.com/1823661/213151588-3740e4e0-d766-4764-9b2a-f0767a44982d.png)

**Comment by franz1981** [comment]
Did you tried outside of docker?
Docker still uses syscall-like banning, that could have some impact (on docker) and/or some skidding (search in the Intel site about what it is) could be in place that can cause blaming it while its nearest inlined methods could be the real guilty ones.

Sadly inlining and Intrinsics often makes profilers confused, including async profiler that indeed relies on AsyncGetCallTrace, that can be prone to point to the wrong method to blame in such cases.

**Comment by jfrantzius** [comment]
It turns out `tsc` is not available as clocksource on my host:
```
$ cat /sys/devices/system/clocksource/clocksource0/available_clocksource
hpet acpi_pm 
```
I found this in /var/log/syslog:
```
tsc: Marking TSC unstable due to clocksource watchdog
TSC found unstable after boot, most likely due to broken BIOS. Use 'tsc=unstable'.
```
Am now trying to find out whether that can be fixed...

**Reply by jfrantzius** [reply]
This could really be a BIOS bug: https://bugzilla.kernel.org/show_bug.cgi?id=202525

The BIOS can only be updated from Windows, but I wiped out the Windows partition, doh!

**Reply by franz1981** [reply]
Doh!

**Comment by jfrantzius** [comment, marked as answer]
For the records, updating the BIOS wasn't enough, I also had to set the Kernel boot parameter `processor.max_cstate=1`.

Now that TSC is available as clocksource, `System.nanoTime()` doesn't show up anymore in profiling.
